components/brave_news/browser/brave_news_pref_manager.cc

```diff
@@ -0,0 +1,378 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_news/browser/brave_news_pref_manager.h"
+
+#include <algorithm>
+#include <iterator>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "base/containers/contains.h"
+#include "base/containers/flat_map.h"
+#include "base/containers/flat_set.h"
+#include "base/functional/bind.h"
+#include "base/ranges/algorithm.h"
+#include "base/uuid.h"
+#include "brave/components/brave_news/browser/brave_news_p3a.h"
+#include "brave/components/brave_news/browser/channel_migrator.h"
+#include "brave/components/brave_news/browser/locales_helper.h"
+#include "brave/components/brave_news/common/brave_news.mojom.h"
+#include "brave/components/brave_news/common/pref_names.h"
+#include "components/prefs/pref_change_registrar.h"
+#include "components/prefs/pref_observer.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+
+namespace brave_news {
+
+bool GetIsEnabled(PrefService* prefs) {
+  bool should_show = prefs->GetBoolean(prefs::kNewTabPageShowToday);
+  bool opted_in = prefs->GetBoolean(prefs::kBraveNewsOptedIn);
+  return should_show && opted_in;
+}
+
+SubscriptionsDiff::SubscriptionsDiff() = default;
+SubscriptionsDiff::~SubscriptionsDiff() = default;
+SubscriptionsDiff& SubscriptionsDiff::operator=(SubscriptionsDiff&&) = default;
+SubscriptionsDiff::SubscriptionsDiff(SubscriptionsDiff&&) = default;
+
+BraveNewsSubscriptions::BraveNewsSubscriptions() = default;
+
+BraveNewsSubscriptions::BraveNewsSubscriptions(
+    base::flat_set<std::string> enabled_publishers,
+    base::flat_set<std::string> disabled_publishers,
+    std::vector<DirectFeed> direct_feeds,
+    base::flat_map<std::string, std::vector<std::string>> channels)
+    : enabled_publishers(std::move(enabled_publishers)),
+      disabled_publishers(std::move(disabled_publishers)),
+      direct_feeds(std::move(direct_feeds)),
+      channels(std::move(channels)) {}
+
+BraveNewsSubscriptions::BraveNewsSubscriptions(const BraveNewsSubscriptions&) =
+    default;
+BraveNewsSubscriptions& BraveNewsSubscriptions::operator=(
+    const BraveNewsSubscriptions&) = default;
+
+BraveNewsSubscriptions::BraveNewsSubscriptions(BraveNewsSubscriptions&&) =
+    default;
+BraveNewsSubscriptions& BraveNewsSubscriptions::operator=(
+    BraveNewsSubscriptions&&) = default;
+
+BraveNewsSubscriptions::~BraveNewsSubscriptions() = default;
+
+std::vector<std::string> BraveNewsSubscriptions::GetChannelLocales() const {
+  std::vector<std::string> locales;
+  for (const auto& [key, value] : channels) {
```

Comment:
checking `DCHECK(!base::Contains(locales, key))` here is worth?

Reply:
Channels is a `flat_map`, so it only ever has one instance of each key, so I don't think we need to worry here.